### PR TITLE
Updating the plugin docs to the V2 Android Plugin APIs

### PR DESCRIPTION
**Pull request by sfshaza2**
Please review! I made a pass through all of the pages in the Packages & Plugins directory. I updated what I could, based on my reading of Chris' blog post and other docs. But please let me know if I've written any fiction!

Also, I have a few PENDING comments and I'd appreciate if your can collectively help me fill those sections in.

Look at all the pages in the Packages & Plugins directory. It's staged here:
https://sz-flutter.firebaseapp.com/docs/development/packages-and-plugins/using-packages

Thanks!

### CODE REVIEW

File: src/docs/development/platform-integration/c-interop.md [hunk 1]
```diff
@@ -350,7 +350,7 @@ in binary form, use the following instructions:
 
 ### Web
 
-Plugins are not yet supported for web apps.
+PENDING
```
Comment (ditman) [inline src/docs/development/platform-integration/c-interop.md]: AFAIK ffi is not supported in Web.

File: src/docs/development/platform-integration/platform-channels.md [hunk 2]
```diff
@@ -4,13 +4,15 @@ short-title: Platform-specific code
 description: Learn how to write custom platform-specific code in your app.
 ---
 
+[PENDING: Is this page up-to-date?]
```
Comment (ditman) [inline src/docs/development/platform-integration/platform-channels.md]: This is still interesting, particularly the part about data serialization and supported types. I think this is still current and valid for all "native" plugins. Only "web in a federated architecture" doesn't use the platform channels.

File: src/docs/development/platform-integration/platform-channels.md [hunk 3]
```diff
@@ -632,47 +630,37 @@ and the app displays 'battery info unavailable'.
...
 
 ## Publish platform-specific code as a package {#publish}
 
-To share your platform-specific code with other developers in the Flutter
-ecosystem, see [publishing
-packages](/docs/development/packages-and-plugins/developing-packages#publish).
+To share your platform-specific code with other developers
+in the Flutter ecosystem, see [publishing packages][].
 
 ## Custom channels and codecs
```
Comment (ditman) [inline src/docs/development/platform-integration/platform-channels.md]: There's an example of a custom codec in the cloud_firestore plugin, that is able to serialize/deserialize many more types than the default ones, here:

https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/method_channel/utils/firestore_message_codec.dart

File: src/docs/cookbook/forms/focus.md [hunk 4]
```diff
@@ -147,7 +147,7 @@ FloatingActionButton(
   While the following code runs successfully in DartPad,
   it fails due to [Issue 52221][] when run against the 
   stable (1.12) or beta (1.14) versions of the Flutter SDK.
-  Until this [fix has landed][] (hash [flutter/flutter@bf551a3])[]
+  Until this [fix has landed][] (hash [flutter/flutter@bf551a3][])
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/cookbook/forms/focus.md]: :D

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 5]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
-* `lib/hello.dart`:
+**`lib/hello.dart`**
 : The Dart code for the package.
-* `test/hello_test.dart`:
-: The [unit tests] for the package.
+
+**`test/hello_test.dart`**
+: The [unit tests][] for the package.
+
+[PENDING: Is this still true for regular Dart packages?]
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: D'OH. Why didn't I think of that! Thanks. My next commit addresses this.

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 6]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
 ### Step 2: Implement the package
 
-For pure Dart packages, simply add the functionality inside the main
-`lib/<package name>.dart` file, or in several files in the `lib` directory.
+For pure Dart packages, simply add the functionality
+inside the main `lib/<package name>.dart` file,
+or in several files in the `lib` directory.
 
-To test the package, add [unit tests]
+To test the package, add [unit tests][]
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: It's our convention. The empty brackets aren't technically necessary, but we were told it's a best practice for some reason I can't remember. So we add them in when we see them.

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 7]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
-*For information about the `flutter.plugin.platforms` key see: [Specifying a plugin's supported platforms](#plugin-platforms).*
+If you want to develop a package that calls into
+platform-specific APIs, you need to develop a plugin package.
+A plugin package is a specialized version of a
+Dart package that, in addition to the content described above,
+also contains platform-specific implementations written for
+Android (Kotlin or Java code), iOS (Swift or Objective-C),
+web (Dart), macos (Dart), or any subset thereof.
+The API is connected to the platform-specific
+implementation(s) using [platform channels][].
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Nice!

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 8]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
+
+**app-facing package**
+: A mandatory package that you include in your Dart code
+  in order to use the plugin. This package specifies the
+  API used by the Flutter app.
+
+**platform package(s)**
+: One or more packages that contain the platform-specific
+  implementation code. The app-facing package calls into
+  these packages&mdash;they aren't included into an app.
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: thx

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 9]
```diff
@@ -225,45 +340,66 @@ It is recommended practice to add the following documentation to all packages:
...
+published to pub.dev/documentation.
+For example, see the docs for [`device_info`][].
 
-If you wish to generate API documentation locally on your development machine, use the following commands:
+If you wish to generate API documentation locally on
+your development machine, use the following commands:
 
-1. Change directory to the location of your package:
+<ol markdown="1">
+<li markdown="1">Change directory to the location of your package:
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: It basically breaks down when you start adding code blocks, images, etc.

File: src/docs/development/platform-integration/c-interop.md (hunk 1 above)
Comment (sfshaza2) [reply src/docs/development/platform-integration/c-interop.md]: Thx

File: src/docs/development/platform-integration/platform-channels.md (hunk 2 above)
Comment (sfshaza2) [reply src/docs/development/platform-integration/platform-channels.md]: Thanks!

File: src/docs/development/platform-integration/platform-channels.md (hunk 3 above)
Comment (sfshaza2) [reply src/docs/development/platform-integration/platform-channels.md]: Excellent!

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 10]
```diff
@@ -4,218 +4,370 @@ short-title: Developing
...
+  thereof.
+  A concrete example is the [`url_launcher`][] plugin package.
+  To see how to use the `url_launcher` package, and how it
+  was extended to implement support for web,
+  see the Medium article by Harry Terkelsen,
+  [How to Write a Flutter Web Plugin, Part 1][].
 
 ## Developing Dart packages {#dart}
 
+The following instructions explain how to write a Dart package
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Very good point. Thanks for catching that.

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 11]
```diff
@@ -4,218 +4,370 @@ short-title: Developing
 description: How to write packages and plugins for Flutter.
 ---
 
+{{site.note.alert}}
+  If you write plugins for Flutter, you should know that
+  the Android plugin API was upgraded to 2.0 in Flutter 1.12
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Done

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 12]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
+another for web, and another for your car
+(as an example of an embedded internet of things device).
+Among other benefits, this approach allows a
+domain expert to extend an existing plugin to work for
+the platform they know best.
+
+A federated plugin requires the following packages:
+
+**app-facing package**
+: A mandatory package that you include in your Dart code
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Done

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 13]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
+  in order to use the plugin. This package specifies the
+  API used by the Flutter app.
+
+**platform package(s)**
+: One or more packages that contain the platform-specific
+  implementation code. The app-facing package calls into
+  these packages&mdash;they aren't included into an app.
+
+**platform interface package**
+: The mandatory package that glues the app-facing packing
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Done

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 14]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
+        pluginClass: HelloPlugin
+      web:
+        pluginClass: HelloPlugin
+        fileName: hello_web.dart
+
+environment:
+  sdk: ">=2.1.0 <3.0.0"
+  # Flutter versions prior to 1.10 did not support the
+  # flutter.plugin.platforms map.
+  flutter: ">=1.10.0 <2.0.0"
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: ok

File: src/docs/development/packages-and-plugins/developing-packages.md [hunk 15]
```diff
@@ -4,218 +4,333 @@ short-title: Developing
...
-```
-flutter:
-  plugin:
-    platforms:
-      android:
-        package: com.example.hello
-        pluginClass: HelloPlugin
-      ios:
-        pluginClass: HelloPlugin
+PENDING
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/docs/development/packages-and-plugins/developing-packages.md]: Done

**Comment by sfshaza2** [comment]
@ditman, @johnpryan, I believe I've addressed all of your feedback. Thanks so much! Any reason that I can't merge this now? (All of the PENDING comments are removed or commented out.)

**Comment by sfshaza2** [comment]
Thanks!

**Pull request merged**